core/src/main/java/tc/oc/pgm/loot/LootableMatchModule.java

```diff
@@ -0,0 +1,302 @@
+package tc.oc.pgm.loot;
+
+import static tc.oc.pgm.util.Assert.assertNotNull;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Random;
+import java.util.function.Predicate;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.bukkit.block.BlockState;
+import org.bukkit.block.Chest;
+import org.bukkit.block.DoubleChest;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.inventory.InventoryOpenEvent;
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.InventoryHolder;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.Nullable;
+import tc.oc.pgm.api.filter.Filter;
+import tc.oc.pgm.api.filter.query.InventoryQuery;
+import tc.oc.pgm.api.filter.query.LocationQuery;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.events.ListenerScope;
+import tc.oc.pgm.filters.FilterMatchModule;
+import tc.oc.pgm.filters.query.BlockQuery;
+import tc.oc.pgm.filters.query.EntityQuery;
+import tc.oc.pgm.itemmeta.ItemModifyMatchModule;
+import tc.oc.pgm.kits.Slot;
+import tc.oc.pgm.util.Pair;
+import tc.oc.pgm.util.collection.InstantMap;
+
+@ListenerScope(value = MatchScope.LOADED)
+public class LootableMatchModule implements MatchModule, Listener {
+
+  private final Logger logger;
+  private final Match match;
+  private final List<FillerDefinition> fillers;
+  private final List<Cache> caches;
+
+  private final InstantMap<Pair<Fillable, FillerDefinition>> filledAt;
+  private final @Nullable ItemModifyMatchModule immm;
+
+  public LootableMatchModule(
+      Logger logger, Match match, List<FillerDefinition> fillers, List<Cache> caches) {
+    this.logger = logger;
+    this.fillers = fillers;
+    this.match = match;
+    this.caches = caches;
+    this.filledAt = new InstantMap<>(new WorldTickClock(match.getWorld()));
+    this.immm = match.getModule(ItemModifyMatchModule.class);
```

Comment:
Consider using moduleOptional here, and instead of having a nullable immm make it an optional. The benefit is you can just imm.ifPresent(...) instead of a null-check, could be more convenient sometimes.

Reply:
I mean its only used a single place in this class, feels like unnecessary object allocation/logic for a simple line looking a little more elegant?